### PR TITLE
Add github Action and fix CMake build fail

### PR DESCRIPTION
**Pull request by bensuperpc**
**Add github Action and fix CMake build fail.**

Tested with manjaro 21.1.6 and [github action](https://github.com/bensuperpc/SDL_mixer/actions/runs/1454991608) (**Ubuntu CI**) in my git repository.

Github action:
- Build with CMake (**Release** mode) and make
- Build With GCC 9 and Clang 10 (ubuntu-latest = ubuntu-20.04 now)
- Build with static or shared lib (CMake only)

Fix wrong header path with SDL2 and bump CMake to 3.0.0 (#314)

Signed-off-by: Bensuperpc <bensuperpc@gmail.com>

### CODE REVIEW

File: include/SDL_mixer.h [hunk 1]
```diff
@@ -22,12 +22,12 @@
 #ifndef SDL_MIXER_H_
 #define SDL_MIXER_H_
 
-#include "SDL_stdinc.h"
-#include "SDL_rwops.h"
-#include "SDL_audio.h"
-#include "SDL_endian.h"
-#include "SDL_version.h"
-#include "begin_code.h"
+#include "SDL2/SDL_stdinc.h"
```
Comment (sezero) [inline include/SDL_mixer.h:25]: The include path changes are incorrect.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.0.0)
```
Comment (sezero) [inline CMakeLists.txt:1]: Why exactly is this necessary?

Comment (bensuperpc) [reply CMakeLists.txt:1]: I no longer have a warning when switching to version 3.0.0, And it's the same version as for the SDL2 lib: https://github.com/libsdl-org/SDL/blob/main/CMakeLists.txt#L5

Comment (sezero) [reply CMakeLists.txt:1]: > I no longer have a warning when switching

You never mentioned any warnings: If you post the warnings you get, it'll help everyone.

File: include/SDL_mixer.h (hunk 1 above)
Comment (bensuperpc) [reply include/SDL_mixer.h:25]: I had errors with header not found, it now works for me and on github CI

Comment (sezero) [reply include/SDL_mixer.h:25]: And your versions will fail with autotools and other build systems.
Those changes are wrong and can't be accepted. The SDL2 discovery
in cmake should be done properly, instead.

File: CMakeLists.txt (hunk 2 above)
Comment (bensuperpc) [reply CMakeLists.txt:1]: > > I no longer have a warning when switching
> 
> You never mentioned any warnings: If you post the warnings you get, it'll help everyone.

I am using CMake 3.21, I cannot get the warning again, i hadn't paid more attention than that, It is perhaps another of my modifications which caused this warning, it is no longer there by reverting to version 2.8.12, I go back to version 2.8.12 ?

File: include/SDL_mixer.h (hunk 1 above)
Comment (bensuperpc) [reply include/SDL_mixer.h:25]: > And your versions will fail with autotools and other build systems. Those changes are wrong and can't be accepted. The SDL2 discovery in cmake should be done properly, instead.

I will change that, do you have the links to these CI ? I would like to know how this is configured, to test in docker containers

File: CMakeLists.txt (hunk 2 above)
Comment (sezero) [reply CMakeLists.txt:1]: Tried myself, `target_sources` is an issue. I'm OK with bumping min version to 3.0.

File: include/SDL_mixer.h (hunk 1 above)
Comment (sezero) [reply include/SDL_mixer.h:25]: The following change to `CMakeLists.txt` works just fine for me:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 283f588..0934778 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.0.0)
 project(SDL_mixer C)
 
 # FIXME: make it able build against system codec libraries, too.
@@ -14,6 +14,9 @@ option(SUPPORT_MP3_MPG123 "Support loading MP3 music via MPG123" OFF)
 option(SUPPORT_MOD_MODPLUG "Support loading MOD music via modplug" OFF)
 option(SUPPORT_MID_TIMIDITY "Support TiMidity" OFF)
 
+find_package(SDL2 REQUIRED)
+include_directories(${SDL2_INCLUDE_DIRS} ${SDL2_INCLUDE_DIR})
+
 include_directories(include src src/codecs)
 
 add_library(SDL2_mixer SHARED)
@@ -64,5 +67,5 @@ if (SUPPORT_MID_TIMIDITY)
 endif()
 
 target_include_directories(SDL2_mixer PUBLIC include)
-target_link_libraries(SDL2_mixer PRIVATE SDL2)
+target_link_libraries(SDL2_mixer PRIVATE ${SDL2_LIBRARIES})
 
```

Comment (sezero) [reply include/SDL_mixer.h:25]: > do you have the links to these CI ?

If the question is how to setup CI to use autotools (configure) or MS Visual Studio or Xcode,
I don't know.

The above patch to CMakeLists.txt is the correct way for SDL2 discovery, include directives
and linkage.

File: CMakeLists.txt [hunk 3]
```diff
@@ -71,5 +67,5 @@ if (SUPPORT_MID_TIMIDITY)
 endif()
 
 target_include_directories(SDL2_mixer PUBLIC include)
-target_link_libraries(SDL2_mixer PRIVATE SDL2)
+target_link_libraries(SDL2_mixer PRIVATE ${SDL2_LIBRARIES})
```
Comment (sezero) [inline CMakeLists.txt:70]: append ` ${SDL2_LIBRARY}` here for better compatibility.

File: include/SDL_mixer.h (hunk 1 above)
Comment (bensuperpc) [reply include/SDL_mixer.h:25]: > The following change to `CMakeLists.txt` works just fine for me:
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 283f588..0934778 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -1,4 +1,4 @@
> -cmake_minimum_required(VERSION 2.8.12)
> +cmake_minimum_required(VERSION 3.0.0)
>  project(SDL_mixer C)
>  
>  # FIXME: make it able build against system codec libraries, too.
> @@ -14,6 +14,9 @@ option(SUPPORT_MP3_MPG123 "Support loading MP3 music via MPG123" OFF)
>  option(SUPPORT_MOD_MODPLUG "Support loading MOD music via modplug" OFF)
>  option(SUPPORT_MID_TIMIDITY "Support TiMidity" OFF)
>  
> +find_package(SDL2 REQUIRED)
> +include_directories(${SDL2_INCLUDE_DIRS} ${SDL2_INCLUDE_DIR})
> +
>  include_directories(include src src/codecs)
>  
>  add_library(SDL2_mixer SHARED)
> @@ -64,5 +67,5 @@ if (SUPPORT_MID_TIMIDITY)
>  endif()
>  
>  target_include_directories(SDL2_mixer PUBLIC include)
> -target_link_libraries(SDL2_mixer PRIVATE SDL2)
> +target_link_libraries(SDL2_mixer PRIVATE ${SDL2_LIBRARIES} ${SDL2_LIBRARY})
>  
> ```

It works for me too, and on CI: https://github.com/bensuperpc/SDL_mixer/actions/runs/1455167527

I have a warning that appeared (**but not on CI**):

```
cmake -S . -B build -G Ninja -DFIND_LOCAL_SDL2=OFF && ninja -C build
-- The C compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    FIND_LOCAL_SDL2


-- Build files have been written to: SDL_mixer/build
ninja: Entering directory `build'
[21/21] Linking C static library libSDL2_mixer.a
```

File: CMakeLists.txt (hunk 3 above)
Comment (sezero) [reply CMakeLists.txt:70]: And also please revert the `#include` path changes and squash the commits for easier review.

File: src/codecs/load_aiff.c [hunk 4]
```diff
@@ -29,6 +29,7 @@
 */
 
 #include "SDL_endian.h"
+
```
Comment (sezero) [inline src/codecs/load_aiff.c]: please revert the changes to all `*.c` files: it _should_ build fine. If not, let's see details why.

**Comment by bensuperpc** [comment]
I think it's good for a review @sezero

**Comment by sezero** [comment]
Looks fair now, finally.

One thing I don't like is the default build is static: can't we make it build shared library by default, instead? (e.g. by setting `BUILD_SHARED_LIBS` to `TRUE` or something..)

**Comment by bensuperpc** [comment]
> Looks fair now, finally.
> 
> One thing I don't like is the default build is static: can't we make it build shared library by default, instead? (e.g. by setting `BUILD_SHARED_LIBS` to `TRUE` or something..)

Like this ?
`option(BUILD_SHARED_LIBS "Enable shared library" ON)`

It is added

**Comment by sezero** [comment]
> Like this ? `option(BUILD_SHARED_LIBS "Enable shared library" ON)`
> 
> It is added

Yes.

The cmake support in SDL_mixer has _a lot of_ ways to go, but this is a fair start
and we can improve upon it later: Merging, thanks!

**Comment by bensuperpc** [comment]
You must activate the CI in the settings of the git repository if it is not activated automatically.

More information: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository

In the next PR, I am thinking of making a README.md with CI badges ^^

**Comment by sezero** [comment]
> You must activate the CI in the settings of the git repository if it is not activated automatically
> 
> More information: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository

@slouken, @icculus: This job is for you guys.

**Comment by slouken** [comment]
Okay, actions have been enabled.
